openverse_catalog/dags/providers/provider_api_scripts/europeana.py

```diff
@@ -8,266 +8,226 @@
+    """
+    Used to decorate RecordBuilder methods for "required" fields
+    to shortcut record building in the case where a record would
+    be missing some required fields and be thrown out anyway.
+    """
 
+    @functools.wraps(fn)
+    def inner(*args, **kwargs):
+        value = fn(*args, **kwargs)
 
-def main(date):
-    logger.info(f"Processing Europeana API for date: {date}")
+        if not value:
+            raise EmptyRequiredFieldException(fn.__name__, value)
 
-    start_timestamp, end_timestamp = _derive_timestamp_pair(date)
-    _get_pagewise(start_timestamp, end_timestamp)
+        return value
 
-    total_images = image_store.commit()
-    logger.info(f"Total images: {total_images}")
-    logger.info("Terminated!")
+    return inner
 
 
-def _get_pagewise(start_timestamp, end_timestamp):
-    cursor = "*"
+class EuropeanaRecordBuilder:
+    """
+    A small class to contain the record building functionality
+    and simplify testing a bit.
+    """
 
-    while cursor is not None:
-        image_list, next_cursor, total_number_of_images = _get_image_list(
-            start_timestamp, end_timestamp, cursor
-        )
+    def get_record_data(self, data: dict) -> dict:
+        try:
+            record = {
+                "foreign_landing_url": self._get_foreign_landing_url(data),
+                "image_url": self._get_image_url(data),
+                "foreign_identifier": self._get_foreign_identifier(data),
+                "meta_data": self._get_meta_data_dict(data),
+                "title": self._get_title(data),
+                "license_info": get_license_info(
+                    license_url=self._get_license_url(data)
+                ),
+            }
+
+            data_providers = set(record["meta_data"]["dataProvider"])
+            eligible_sub_providers = {
+                s
+                for s in EuropeanaDataIngester.sub_providers
+                if EuropeanaDataIngester.sub_providers[s] in data_providers
+            }
+            if len(eligible_sub_providers) > 1:
+                raise Exception(
+                    f"More than one sub-provider identified for the "
+                    f"image with foreign ID {record['foreign_identifier']}"
+                )
+
+            return record | {
+                "source": (
+                    eligible_sub_providers.pop()
+                    if len(eligible_sub_providers) == 1
+                    else EuropeanaDataIngester.providers["image"]
+                )
+            }
+        except EmptyRequiredFieldException as exc:
+            logger.warning("A required field was empty", exc_info=exc)
+            return None
+
+    @raise_if_empty
+    def _get_image_url(self, data: dict) -> str | None:
+        group = data.get("edmIsShownBy")
+        return group[0] if group else None
+
+    @raise_if_empty
+    def _get_foreign_identifier(self, data: dict) -> str | None:
+        return data.get("id")
+
+    @raise_if_empty
```

Comment:
A title is not required for a media item, some of them have `""` as the title. This might not be needed for Europeana, but is necessary for some providers who have user-provided items (such as Flickr).

Reply:
Agreed -- I'm not _sure_ if it's absolutely necessary to remove this check for Europeana or if we can assume all records will have a title, but since this will halt ingestion entirely if it raises I'd prefer to remove this check.